File: plugins/destination/sqlite/client/migrate.go
package client

import (
	"context"
	"fmt"
	"strings"

	"github.com/apache/arrow-go/v18/arrow"
	"github.com/cloudquery/plugin-sdk/v4/message"
	"github.com/cloudquery/plugin-sdk/v4/schema"
)

const (
	sqlTableInfo = "PRAGMA table_info('%s');"
)

type columnInfo struct {
	index        int
	name         string
	typ          string
	notNull      bool
	defaultValue any
	pk           int
}

type tableInfo struct {
	columns []columnInfo
}

func identifier(str string) string {
	return `"` + str + `"`
}

func (c *Client) sqliteTables(tables schema.Tables) (schema.Tables, error) {
	var schemaTables schema.Tables
	for _, table := range tables {
		var columns []schema.Column
		info, err := c.getTableInfo(table.Name)
		if info == nil {
			continue
		}
		if err != nil {
			return nil, err
		}
		for _, col := range info.columns {
			columns = append(columns, schema.Column{
				Name:       col.name,
				Type:       c.sqliteTypeToArrowType(col.typ),
				PrimaryKey: col.pk != 0,
				NotNull:    col.notNull,
			})
		}
		schemaTables = append(schemaTables, &schema.Table{Name: table.Name, Columns: columns})
	}
	return schemaTables, nil
}

func (c *Client) normalizeTables(tables schema.Tables) schema.Tables {
	flattened := tables.FlattenTables()
	normalized := make(schema.Tables, len(flattened))
	for i, table := range flattened {
		normalized[i] = c.normalizeTable(table)
	}
	return normalized
}

func (c *Client) normalizeTable(table *schema.Table) *schema.Table {
	columns := make([]schema.Column, len(table.Columns))
	for i, col := range table.Columns {
		normalized := c.normalizeField(col.ToArrowField())
		columns[i] = schema.NewColumnFromArrowField(*normalized)
	}
	return &schema.Table{Name: table.Name, Columns: columns}
}

func (c *Client) normalizeField(field arrow.Field) *arrow.Field {
	return &arrow.Field{
		Name:     field.Name,
		Type:     c.arrowTypeToSqlite(field.Type),
		Nullable: field.Nullable,
		Metadata: field.Metadata,
	}
}

func (c *Client) nonAutoMigratableTables(tables schema.Tables, sqliteTables schema.Tables, safeTables map[string]bool) map[string][]schema.TableColumnChange {
	result := make(map[string][]schema.TableColumnChange)
	for _, t := range tables {
		sqliteTable := sqliteTables.Get(t.Name)
		if sqliteTable == nil {
			continue
		}
		changes := sqliteTable.GetChanges(t)
		if safeTables[t.Name] && !c.canAutoMigrate(changes) {
			result[t.Name] = changes
		}
	}
	return result
}

func (c *Client) autoMigrateTable(ctx context.Context, table *schema.Table, changes []schema.TableColumnChange) error {
	for _, change := range changes {
		if change.Type == schema.TableColumnChangeTypeAdd {
			if err := c.addColumn(ctx, table.Name, change.Current.Name, c.arrowTypeToSqliteStr(change.Current.Type)); err != nil {
				return err
			}
		}
	}
	return nil
}

func (*Client) canAutoMigrate(changes []schema.TableColumnChange) bool {
	for _, change := range changes {
		switch change.Type {
		case schema.TableColumnChangeTypeAdd:
			if change.Current.PrimaryKey || change.Current.NotNull {
				return false
			}
		case schema.TableColumnChangeTypeRemove:
			if change.Previous.PrimaryKey || change.Previous.NotNull {
				return false
			}
		case schema.TableColumnChangeTypeRemoveUniqueConstraint:
			continue
		default:
			return false
		}
	}
	return true
}

// This is the responsibility of the CLI of the client to lock before running migration
func (c *Client) MigrateTables(ctx context.Context, msgs message.WriteMigrateTables) error {
	allTables := schema.Tables{}
	safeTables := make(map[string]bool)
	for _, msg := range msgs {
		allTables = append(allTables, msg.Table)
		safeTables[msg.Table.Name] = !msg.MigrateForce
	}

	normalizedTables := c.normalizeTables(allTables)
	sqliteTables, err := c.sqliteTables(normalizedTables)
	if err != nil {
		return err
	}

	nonAutoMigratableTables := c.nonAutoMigratableTables(normalizedTables, sqliteTables, safeTables)
	if len(nonAutoMigratableTables) > 0 {
		return fmt.Errorf("\nCan't migrate tables automatically, migrate manually or consider using 'migrate_mode: forced'. Non auto migratable tables changes:\n\n%s", schema.GetChangesSummary(nonAutoMigratableTables))
	}

	for _, table := range normalizedTables {
		c.logger.Info().Str("table", table.Name).Msg("Migrating table")
		if len(table.Columns) == 0 {
			c.logger.Info().Str("table", table.Name).Msg("Table with no columns, skipping")
			continue
		}

		sqlite := sqliteTables.Get(table.Name)
		if sqlite == nil {
			c.logger.Debug().Str("table", table.Name).Msg("Table doesn't exist, creating")
			if err := c.createTableIfNotExist(table); err != nil {
				return err
			}
		} else {
			changes := table.GetChanges(sqlite)
			if c.canAutoMigrate(changes) {
				c.logger.Info().Str("table", table.Name).Msg("Table exists, auto-migrating")
				if err := c.autoMigrateTable(ctx, table, changes); err != nil {
					return err
				}
			} else {
				c.logger.Info().Str("table", table.Name).Msg("Table exists, force migration required")
				if err := c.recreateTable(ctx, table); err != nil {
					return err
				}
			}
		}
	}

	return nil
}

func (c *Client) recreateTable(ctx context.Context, table *schema.Table) error {
	sql := "drop table if exists " + identifier(table.Name)
	if _, err := c.db.ExecContext(ctx, sql); err != nil {
		return fmt.Errorf("failed to drop table %s: %w", table.Name, err)
	}
	return c.createTableIfNotExist(table)
}

func (c *Client) addColumn(ctx context.Context, tableName string, columnName string, columnType string) error {
	sql := "alter table " + identifier(tableName) + " add column " + identifier(columnName) + " " + identifier(columnType)
	if _, err := c.db.ExecContext(ctx, sql); err != nil {
		return fmt.Errorf("failed to add column %s on table %s: %w", columnName, tableName, err)
	}
	return nil
}

func (c *Client) createTableIfNotExist(table *schema.Table) error {
	var sb strings.Builder

	sb.WriteString("CREATE TABLE IF NOT EXISTS ")
	sb.WriteString(identifier(table.Name))
	sb.WriteString(" (")
	totalColumns := len(table.Columns)

	primaryKeys := []string{}
	for i, col := range table.Columns {
		sqlType := c.arrowTypeToSqliteStr(col.Type)
		if sqlType == "" {
			c.logger.Warn().Str("table", table.Name).Str("column", col.Name).Msg("Column type is not supported, skipping")
			continue
		}
		fieldDef := identifier(col.Name) + ` ` + sqlType
		if col.NotNull {
			fieldDef += " NOT NULL"
		}
		sb.WriteString(fieldDef)
		if i != totalColumns-1 {
			sb.WriteString(",")
		}
		if col.PrimaryKey {
			primaryKeys = append(primaryKeys, identifier(col.Name))
		}
	}

	if len(primaryKeys) > 0 {
		// add composite PK constraint on primary key columns
		sb.WriteString(", CONSTRAINT ")
		sb.WriteString(identifier(table.Name + "_cqpk"))
		sb.WriteString(" PRIMARY KEY (")
		sb.WriteString(strings.Join(primaryKeys, ","))
		sb.WriteString(")")
	}
	sb.WriteString(")")
	_, err := c.db.Exec(sb.String())
	if err != nil {
		return fmt.Errorf("failed to create table with '%s': %w", sb.String(), err)
	}
	return nil
}

func (c *Client) getTableInfo(tableName string) (*tableInfo, error) {
	info := tableInfo{}
	rows, err := c.db.Query(fmt.Sprintf(sqlTableInfo, tableName))
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		colInfo := columnInfo{}
		if err := rows.Scan(
			&colInfo.index,
			&colInfo.name,
			&colInfo.typ,
			&colInfo.notNull,
			&colInfo.defaultValue,
			&colInfo.pk); err != nil {
			return nil, err
		}
		colInfo.typ = strings.ToLower(colInfo.typ)
		info.columns = append(info.columns, colInfo)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	if len(info.columns) == 0 {
		// Table doesn't exist
		return nil, nil
	}
	return &info, nil
}

File: plugins/destination/sqlite-python/plugin/sqlite/migrate.py
import sqlite3
from typing import List, Any, Tuple, Optional
from cloudquery.sdk.schema import Table, Column, TableColumnChange, TableColumnChangeType, get_table_changes, flatten_tables
import pyarrow as pa
from cloudquery.sdk.message import WriteMigrateTableMessage
from plugin.sqlite.type_conversions import arrow_type_to_sqlite_str, arrow_type_to_sqlite, sqlite_type_to_arrow_type

SQL_TABLE_INFO = "PRAGMA table_info('{}');"

class ColumnInfo:
    def __init__(self, index: int, name: str, typ: str, not_null: bool, default_value: Any, pk: int):
        self.index = index
        self.name = name
        self.typ = typ.lower()
        self.not_null = not_null
        self.default_value = default_value
        self.pk = pk

class TableInfo:
    def __init__(self):
        self.columns: List[ColumnInfo] = []

class MigrateSQLClient:
    def __init__(self, db: sqlite3.Connection):
        self.db = db

    def identifier(self, string: str) -> str:
        return f'"{string}"'

    def sqlite_tables(self, tables: List[Table]) -> List[Table]:
        schema_tables = []
        for table in tables:
            columns = []
            info = self.get_table_info(table.name)
            if info is None:
                continue
            for col in info.columns:
                columns.append(Column(
                    name=col.name,
                    type=sqlite_type_to_arrow_type(col.typ),
                    primary_key=col.pk != 0,
                    not_null=col.not_null
                ))
            schema_tables.append(Table(name=table.name, columns=columns))
        return schema_tables

    def normalize_tables(self, tables: List[Table]) -> List[Table]:
        flattened = flatten_tables(tables)
        normalized = [self.normalize_table(table) for table in flattened]
        return normalized

    def normalize_table(self, table: Table) -> Table:
        columns = [self.normalize_field(col.to_arrow_field()) for col in table.columns]
        columns = [Column.from_arrow_field(col) for col in columns]
        return Table(name=table.name, columns=columns)

    def normalize_field(self, field: pa.Field) -> pa.Field:
        return pa.field(
            name=field.name,
            type=arrow_type_to_sqlite(field.type),
            nullable=field.nullable,
            metadata=field.metadata
        )

    def non_auto_migratable_tables(self, tables: List[Table], sqlite_tables: List[Table]) -> Tuple[List[str], List[List[TableColumnChange]]]:
        result = []
        table_changes = []
        for t in tables:
            sqlite_table = self.get_table_by_name(sqlite_tables, t.name)
            if sqlite_table is None:
                continue
            changes = get_table_changes(sqlite_table, t)
            if not self.can_auto_migrate(changes):
                result.append(t.name)
                table_changes.append(changes)
        return result, table_changes

    def auto_migrate_table(self, table: Table, changes: List[TableColumnChange]) -> Optional[Exception]:
        for change in changes:
            if change.type == TableColumnChangeType.ADD:
                err = self.add_column(table.name, change.current.name, arrow_type_to_sqlite_str(change.current.type))
                if err:
                    return err
        return None

    def can_auto_migrate(self, changes: List[TableColumnChange]) -> bool:
        for change in changes:
            if change.type == TableColumnChangeType.ADD:
                if change.current.primary_key or change.current.not_null:
                    return False
            elif change.type == TableColumnChangeType.REMOVE:
                if change.previous.primary_key or change.previous.not_null:
                    return False
            elif change.type != TableColumnChangeType.REMOVE_UNIQUE_CONSTRAINT:
                return False
        return True

    def migrate_tables(self, msgs: List[WriteMigrateTableMessage]):
        for msg in msgs:
            force = msg.migrate_force
            tables = [msg.table]
            normalized_tables = self.normalize_tables(tables)
            sqlite_tables = self.sqlite_tables(normalized_tables)

            if not force:
                non_auto_migratable_tables, changes = self.non_auto_migratable_tables(normalized_tables, sqlite_tables)
                if non_auto_migratable_tables:
                    return Exception(f"Tables {', '.join(non_auto_migratable_tables)} with changes {changes} require migration. Migrate manually or consider using 'migrate_mode: forced'")

            for table in normalized_tables:
                if len(table.columns) == 0:
                    continue

                sqlite_table = self.get_table_by_name(sqlite_tables, table.name)
                if sqlite_table is None:
                    err = self.create_table_if_not_exist(table)
                    if err:
                        raise err
                else:
                    changes = get_table_changes(table, sqlite_table)
                    if self.can_auto_migrate(changes):
                        err = self.auto_migrate_table(table, changes)
                        if err:
                            raise err
                    else:
                        err = self.recreate_table(table)
                        if err:
                            raise err
        return None

    def recreate_table(self, table: Table) -> Optional[Exception]:
        sql = f"DROP TABLE IF EXISTS {self.identifier(table.name)}"
        try:
            self.db.execute(sql)
        except sqlite3.Error as err:
            return Exception(f"Failed to drop table {table.name}: {err}")
        return self.create_table_if_not_exist(table)

    def add_column(self, table_name: str, column_name: str, column_type: str) -> Optional[Exception]:
        sql = f"ALTER TABLE {self.identifier(table_name)} ADD COLUMN {self.identifier(column_name)} {self.identifier(column_type)}"
        try:
            self.db.execute(sql)
        except sqlite3.Error as err:
            return Exception(f"Failed to add column {column_name} on table {table_name}: {err}")
        return None

    def create_table_if_not_exist(self, table: Table) -> Optional[Exception]:
        sb = []

        sb.append(f"CREATE TABLE IF NOT EXISTS {self.identifier(table.name)} (")
        primary_keys = []
        for i, col in enumerate(table.columns):
            sql_type = arrow_type_to_sqlite_str(col.type)
            if not sql_type:
                continue
            field_def = f"{self.identifier(col.name)} {sql_type}"
            if col.not_null:
                field_def += " NOT NULL"
            sb.append(field_def)
            if col.primary_key:
                primary_keys.append(self.identifier(col.name))
            if i < len(table.columns) - 1:
                sb.append(", ")

        if primary_keys:
            sb.append(f", CONSTRAINT {self.identifier(table.name + '_cqpk')} PRIMARY KEY ({', '.join(primary_keys)})")
        sb.append(")")

        try:
            self.db.execute("".join(sb))
        except sqlite3.Error as err:
            return Exception(f"Failed to create table with '{''.join(sb)}': {err}")
        return None

    def get_table_info(self, table_name: str) -> Optional[TableInfo]:
        info = TableInfo()
        cursor = self.db.cursor()
        try:
            cursor.execute(SQL_TABLE_INFO.format(table_name))
            for row in cursor.fetchall():
                col_info = ColumnInfo(*row)
                info.columns.append(col_info)
        except sqlite3.Error as err:
            raise err
        finally:
            cursor.close()

        if len(info.columns) == 0:
            return None

        return info

    def get_table_by_name(self, tables: List[Table], name: str) -> Optional[Table]:
        for table in tables:
            if table.name == name:
                return table
        return None

